include/deal.II/base/mpi.h

```diff
@@ -1538,6 +1539,85 @@ namespace Utilities
 #  endif
     }
 
+    namespace internal
+    {
+      template <typename T>
+      ArrayView<T>
+      builtin_broadcast(const MPI_Comm &    comm,
+                        const ArrayView<T> &object_to_send,
+                        const unsigned int  root_process)
+      {
+#  ifndef DEAL_II_WITH_MPI
+        (void)comm;
+        (void)root_process;
+        return object_to_send;
+#  else
+        MPI_Datatype datatype =
+          dealii::Utilities::MPI::internal::mpi_type_id(object_to_send);
+
+        ArrayView<T> buffer;
+        if (this_mpi_process(comm) == root_process)
+          {
+            buffer = object_to_send;
+          }
+
+        int ierr =
+          MPI_Bcast(&buffer, buffer.size(), datatype, root_process, comm);
+        AssertThrowMPI(ierr);
+
+        return buffer;
+#  endif
+      }
+
+      template <typename T>
+      T
+      builtin_single_broadcast(const MPI_Comm &   comm,
+                               const T &          object_to_send,
+                               const unsigned int root_process)
+      {
+#  ifndef DEAL_II_WITH_MPI
+        (void)comm;
+        (void)root_process;
+        return object_to_send;
+#  else
+        MPI_Datatype datatype =
+          dealii::Utilities::MPI::internal::mpi_type_id(object_to_send);
+
+        T buffer;
+        if (this_mpi_process(comm) == root_process)
+          {
+            buffer = object_to_send;
+          }
+
+        int ierr = MPI_Bcast(&buffer, 1, datatype, root_process, comm);
+        AssertThrowMPI(ierr);
+
+        return buffer;
+#  endif
+      }
+
+    } // namespace internal
+
+    template <typename T>
+    std::enable_if_t<is_mpi_type<T>::value, T>
+    broadcast(const MPI_Comm &   comm,
+              const T &          object_to_send,
+              const unsigned int root_process)
+    {
+      return internal::builtin_single_broadcast(comm,
+                                                object_to_send,
+                                                root_process);
+    }
+
+
+    template <typename T>
+    std::enable_if_t<is_mpi_type<T>::value, T>
+    broadcast(const MPI_Comm &    comm,
+              const ArrayView<T> &object_to_send,
+              const unsigned int  root_process)
+    {
+      return internal::builtin_broadcast(comm, object_to_send, root_process);
```

Comment:
This will not work, for the reason mentioned above. But, furthermore, the function takes an *array* of `T`s as input but returns only a single `T` as output. That can't be what you had in mind.

I know what you wanted to achieve with this function: If someone passes in an array (or vector, or `std::array`, ...) of objects, then you want to also optimize the broadcast. I know there was a comment that you should use `ArrayView` for that, but the problem is that `ArrayView` can't be the return type.

My suggestion is to remove this function for a moment. Then you just have the single broadcast case above, for which you can move the internal function back into the implementation. Let's do that in a first step and come back to the array case in a second PR. I think that would make it easiest to move forward.

Reply:
Thank you, I applied your advice in 147b71fd64585107f0780dbb0826814473a6d371.